--- a/src/drv_mei_cpe_api_atm_ptm_intern.c
+++ b/src/drv_mei_cpe_api_atm_ptm_intern.c
@@ -321,6 +321,9 @@ IFX_int32_t MEI_InternalTcRequest(
    /* get bonding flag */
    is_bonding = pArgTcRequest->is_bonding;
 
+   g_tx_link_rate[dslLineNum][0] = 0;
+   g_tx_link_rate[dslLineNum][1] = 0;
+
    /* get NULL or function pointer */
    mei_tc_request =
       (mei_tc_request_t)ppa_callback_get(LTQ_MEI_TC_REQUEST);
