diff -ru a/drivers/tty/serial/8250/8250_early.c serial/drivers/tty/serial/8250/8250_early.c
--- a/drivers/tty/serial/8250/8250_early.c	2022-05-25 09:18:02.000000000 +0200
+++ serial/drivers/tty/serial/8250/8250_early.c	2022-09-16 05:02:51.327657602 +0200
@@ -109,6 +109,34 @@
 	uart_console_write(port, s, count, serial_putc);
 }
 
+static bool __early_serial8250_write(struct console *con, struct cons_write_context *wctxt,
+				     unsigned char c)
+{
+	struct earlycon_device *device = con->data;
+	struct uart_port *port = &device->port;
+	unsigned char *s = wctxt->outbuf;
+
+	serial_putc(port, c);
+
+	for (; wctxt->pos < wctxt->len; wctxt->pos++, s++) {
+		if (!console_can_proceed(wctxt))
+			return false;
+
+		uart_console_write(port, s, 1, serial_putc);
+	}
+	return true;
+}
+
+static bool early_serial8250_write_thread(struct console *con, struct cons_write_context *wctxt)
+{
+	return __early_serial8250_write(con, wctxt, 'T');
+}
+
+static bool early_serial8250_write_atomic(struct console *con, struct cons_write_context *wctxt)
+{
+	return __early_serial8250_write(con, wctxt, 'A');
+}
+
 #ifdef CONFIG_CONSOLE_POLL
 static int early_serial8250_read(struct console *console,
 				 char *s, unsigned int count)
@@ -160,6 +188,8 @@
 	if (!(device->port.membase || device->port.iobase))
 		return -ENODEV;
 
+	device->port.regshift = 2;
+
 	if (!device->baud) {
 		struct uart_port *port = &device->port;
 		unsigned int ier;
@@ -172,6 +202,10 @@
 
 	device->con->write = early_serial8250_write;
 	device->con->read = early_serial8250_read;
+	device->con->flags &= ~CON_BOOT;
+	device->con->flags |= CON_NO_BKL;
+	device->con->write_thread = early_serial8250_write_thread;
+	device->con->write_atomic = early_serial8250_write_atomic;
 	return 0;
 }
 EARLYCON_DECLARE(uart8250, early_serial8250_setup);
diff -ru a/drivers/tty/serial/8250/Kconfig serial/drivers/tty/serial/8250/Kconfig
--- a/drivers/tty/serial/8250/Kconfig	2022-05-25 09:18:02.000000000 +0200
+++ serial/drivers/tty/serial/8250/Kconfig	2022-09-16 03:47:10.734146829 +0200
@@ -83,9 +83,18 @@
 
 	  If unsure, say N.
 
+config SERIAL_8250_CONSOLE_EARLY
+	bool "Console on 8250/16550 and compatible noBKL console mockup"
+	default SERIAL_8250
+	select SERIAL_CORE_CONSOLE
+	select SERIAL_EARLYCON
+	help
+	  Mockup to demonstrate the core capabilities for noBKL consoles.
+	  OTOH, the _only_ reliable reentrant and NMI safe console...
+
 config SERIAL_8250_CONSOLE
 	bool "Console on 8250/16550 and compatible serial port"
-	depends on SERIAL_8250=y
+	depends on SERIAL_8250=y && !SERIAL_8250_CONSOLE_EARLY
 	select SERIAL_CORE_CONSOLE
 	select SERIAL_EARLYCON
 	help
diff -ru a/drivers/tty/serial/8250/Makefile serial/drivers/tty/serial/8250/Makefile
--- a/drivers/tty/serial/8250/Makefile	2022-05-25 09:18:02.000000000 +0200
+++ serial/drivers/tty/serial/8250/Makefile	2022-09-16 03:47:10.734146829 +0200
@@ -18,7 +18,7 @@
 obj-$(CONFIG_SERIAL_8250_ACORN)		+= 8250_acorn.o
 obj-$(CONFIG_SERIAL_8250_ASPEED_VUART)	+= 8250_aspeed_vuart.o
 obj-$(CONFIG_SERIAL_8250_BCM2835AUX)	+= 8250_bcm2835aux.o
-obj-$(CONFIG_SERIAL_8250_CONSOLE)	+= 8250_early.o
+obj-$(CONFIG_SERIAL_8250_CONSOLE_EARLY)	+= 8250_early.o
 obj-$(CONFIG_SERIAL_8250_FOURPORT)	+= 8250_fourport.o
 obj-$(CONFIG_SERIAL_8250_ACCENT)	+= 8250_accent.o
 obj-$(CONFIG_SERIAL_8250_BOCA)		+= 8250_boca.o
